src/stream.rs

```diff
@@ -11,44 +11,57 @@ use std::{
 
 use tokio::io::{AsyncRead, AsyncWrite, ReadBuf};
 
-/// Stream, either plain TCP or TLS.
+/// A stream that might be protected with TLS.
 #[pin_project(project = StreamProj)]
-pub enum Stream<S, T> {
+pub enum MaybeTlsStream<S> {
     /// Unencrypted socket stream.
     Plain(#[pin] S),
-    /// Encrypted socket stream.
-    Tls(#[pin] T),
+    /// Encrypted socket stream using [`native-tls`].
+    #[cfg(feature = "native-tls")]
+    NativeTls(tokio_native_tls::TlsStream<S>),
+    /// Encrypted socket stream using [`rustls`].
+    #[cfg(feature = "rustls-tls")]
+    Rustls(tokio_rustls::client::TlsStream<S>),
 }
 
-impl<S: AsyncRead + Unpin, T: AsyncRead + Unpin> AsyncRead for Stream<S, T> {
+impl<S: AsyncRead + AsyncWrite + Unpin> AsyncRead for MaybeTlsStream<S> {
```

Comment:
Just curious: is AsyncWrite really needed for implementing AsyncRead? (converse case below)

Reply:
Unfortunately yes, due to `rustls`. Here is the signature of rustls's `TlsSTream`:
```rust
impl<IO> AsyncRead for TlsStream<IO>
where
    IO: AsyncRead + AsyncWrite + Unpin
```

Reply:
I see. Not a real problem.